File: app/src/main/java/com/example/vintageradioapp/ui/VideoPlayerViewModel.kt
package com.example.vintageradioapp.ui

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.example.vintageradioapp.data.SongParser
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.CoroutineDispatcher
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import javax.inject.Inject

@HiltViewModel
class VideoPlayerViewModel @Inject constructor(
    private val songParser: SongParser,
    private val dispatcher: CoroutineDispatcher
) : ViewModel() {
    private val _state = MutableStateFlow(VideoPlayerState())
    val state: StateFlow<VideoPlayerState> = _state.asStateFlow()

    init {
        loadSongs("70")
    }

    fun checkIfIsTabletAndLandscape(isTabletAndLandscape: Boolean) {
        _state.update { it.copy(isTabletAndLandscape = isTabletAndLandscape) }
    }

    fun loadSongs(decade: String) {
        viewModelScope.launch(dispatcher) {
            _state.update {
                it.copy(
                    isLoading = true
                )
            }
            try {
                val songs = songParser.parseSongs(decade).shuffled()
                if (songs.isNotEmpty()) {
                    _state.update {
                        it.copy(
                            songs = songs,
                            isLoading = false,
                            currentSongIndex = 0, // Start with the first song of the shuffled list
                            currentPlaybackTimeSeconds = 0,
                            totalDurationSeconds = 0, // Will be updated by player
                            error = null
                        )
                    }
                } else {
                    _state.update {
                        it.copy(
                            isLoading = false,
                            error = "No songs found for decade $decade in ids.txt."
                        )
                    }
                }
            } catch (e: Exception) {
                _state.update {
                    it.copy(
                        isLoading = false,
                        error = "Error loading songs: ${e.message}"
                    )
                }
            }
        }
    }

    fun onAction(action: VideoPlayerAction) {
        when (action) {
            is VideoPlayerAction.PlayPause -> {
                if (state.value.songs.isEmpty()) return // No action if no songs
                // If already playing, pause. If paused, play.
                val newIsPlaying = !state.value.isPlaying
                _state.update { it.copy(isPlaying = newIsPlaying) }
            }
            is VideoPlayerAction.SetPlaying -> {
                if (state.value.songs.isEmpty()) return
                if (state.value.isPlaying != action.playing) {
                    _state.update { it.copy(isPlaying = action.playing) }
                }
            }
            is VideoPlayerAction.NextSong -> {
                if (state.value.songs.isEmpty()) return
                goToNextSong()
            }
            is VideoPlayerAction.PreviousSong -> {
                if (state.value.songs.isEmpty()) return
                goToPreviousSong()
            }
            is VideoPlayerAction.SeekTo -> { // User finished seeking with slider
                if (state.value.songs.isEmpty()) return
                _state.update {
                    it.copy(
                        currentPlaybackTimeSeconds = action.positionSeconds
                    )
                }
            }
            is VideoPlayerAction.UpdatePlaybackTime -> { // Player reported time
                 if (state.value.songs.isEmpty()) return
                _state.update {
                    it.copy(
                        currentPlaybackTimeSeconds = action.timeSeconds
                    )
                }
            }
            is VideoPlayerAction.UpdateTotalDuration -> {
                 if (state.value.songs.isEmpty()) return
                _state.update {
                    it.copy(
                        totalDurationSeconds = action.durationSeconds
                    )
                }
            }
            is VideoPlayerAction.OnError -> {
                _state.update {
                    it.copy(
                        error = action.error,
                        isLoading = false,
                        isPlaying = false
                    )
                }
            }
            is VideoPlayerAction.DismissError -> {
                goToNextSong()
            }
            is VideoPlayerAction.ChangeDecade -> {
                handleChangeDecade(action.decade)
            }
        }
    }

    private fun goToPreviousSong() {
        _state.update { currentState ->
            val prevIndex = (currentState.currentSongIndex - 1 + currentState.songs.size) %
                currentState.songs.size
            currentState.copy(
                currentSongIndex = prevIndex,
                currentPlaybackTimeSeconds = 0, // Reset time for new song
                totalDurationSeconds = 0, // Reset duration for new song
                isPlaying = true, // Auto-play previous song
                isPrevButtonEnabled = true
            )
        }
    }

    private fun handleChangeDecade(decade: String) {
        _state.update { it.copy(currentDecade = decade) }
        loadSongs(decade)
        _state.update { it.copy(isPrevButtonEnabled = false) }
    }

    private fun goToNextSong() {
        _state.update { currentState ->
            val nextIndex = (currentState.currentSongIndex + 1) % currentState.songs.size
            currentState.copy(
                currentSongIndex = nextIndex,
                currentPlaybackTimeSeconds = 0, // Reset time for new song
                totalDurationSeconds = 0, // Reset duration for new song
                isPlaying = true, // Auto-play next song
                error = null,
                isPrevButtonEnabled = true
            )
        }
    }
}

File: VintageRadioApp/VintageRadioApp/ui/VideoPlayerViewModel.swift
import Foundation
import Combine

enum PlayerCommand {
    case load(videoID: String)
    case play
    case pause
    case seek(to: Double)
}

class VideoPlayerViewModel: ObservableObject {
    @Published private(set) var state = VideoPlayerState()
    let commandPublisher = PassthroughSubject<PlayerCommand, Never>()

    private let songParser = SongParser()

    init() {
        loadSongs(decade: state.currentDecade)
    }

    func onAction(_ action: VideoPlayerAction) {
        switch action {
        case .playPause:
            if state.songs.isEmpty { return }
            state.isPlaying.toggle()
            if state.isPlaying {
                commandPublisher.send(.play)
            } else {
                commandPublisher.send(.pause)
            }
        case .setPlaying(let playing):
            if state.songs.isEmpty { return }
            if state.isPlaying != playing {
                state.isPlaying = playing
            }
        case .nextSong:
            if state.songs.isEmpty { return }
            goToNextSong()
        case .previousSong:
            if state.songs.isEmpty { return }
            goToPreviousSong()
        case .seekTo(let positionSeconds):
            if state.songs.isEmpty { return }
            state.currentPlaybackTimeSeconds = positionSeconds
            commandPublisher.send(.seek(to: positionSeconds))
        case .updatePlaybackTime(let timeSeconds):
            if state.songs.isEmpty { return }
            state.currentPlaybackTimeSeconds = timeSeconds
        case .updateTotalDuration(let durationSeconds):
            if state.songs.isEmpty { return }
            state.totalDurationSeconds = durationSeconds
        case .onPlayerError(let errorCode):
            handlePlayerError(errorCode)
        case .onError(let error):
            state.error = error
            state.isLoading = false
            state.isPlaying = false
        case .dismissError:
            goToNextSong()
        case .changeDecade(let decade):
            handleChangeDecade(decade: decade)
        }
    }

    private func loadSongs(decade: String) {
        state.isLoading = true
        songParser.parseSongs(decade: decade) { [weak self] songs in
            guard let self = self else { return }

            let shuffledSongs = songs.shuffled()
            if !shuffledSongs.isEmpty {
                var newState = self.state
                newState.songs = shuffledSongs
                newState.isLoading = false
                newState.currentSongIndex = 0
                newState.currentPlaybackTimeSeconds = 0
                newState.totalDurationSeconds = 0
                newState.error = nil
                newState.isPlaying = true
                self.state = newState
                if let firstSong = shuffledSongs.first {
                    commandPublisher.send(.load(videoID: firstSong.youtubeID))
                }
            } else {
                var newState = self.state
                newState.isLoading = false
                newState.error = "No songs found for decade \(decade) in ids.txt."
                self.state = newState
            }
        }
    }

    private func goToNextSong() {
        let nextIndex = (state.currentSongIndex + 1) % state.songs.count
        state.currentSongIndex = nextIndex
        state.currentPlaybackTimeSeconds = 0
        state.totalDurationSeconds = 0
        state.isPlaying = true
        state.error = nil
        state.isPrevButtonEnabled = true
        if let newSong = state.currentSong {
            commandPublisher.send(.load(videoID: newSong.youtubeID))
        }
    }

    private func goToPreviousSong() {
        let prevIndex = (state.currentSongIndex - 1 + state.songs.count) % state.songs.count
        state.currentSongIndex = prevIndex
        state.currentPlaybackTimeSeconds = 0
        state.totalDurationSeconds = 0
        state.isPlaying = true
        state.isPrevButtonEnabled = true
        if let newSong = state.currentSong {
            commandPublisher.send(.load(videoID: newSong.youtubeID))
        }
    }

    private func handleChangeDecade(decade: String) {
        state.currentDecade = decade
        loadSongs(decade: decade)
        state.isPrevButtonEnabled = false
    }

    private func handlePlayerError(_ errorCode: Int) {
        let errorMessage: String
        switch errorCode {
        case 2:
            errorMessage = "Player Error: Invalid video ID."
        case 5:
            errorMessage = "Player Error: HTML5 Player issue."
        case 100:
            errorMessage = "Video not found or private."
        case 101, 150:
            errorMessage = "Playback restricted by owner."
        default:
            errorMessage = "An unknown player error occurred."
        }
        state.error = errorMessage

        // Skip to the next song after a short delay
        DispatchQueue.main.asyncAfter(deadline: .now() + 3) {
            if self.state.error == errorMessage { // Only skip if the error is still the same
                self.goToNextSong()
            }
        }
    }
}
